moodler/upload.py

```diff
@@ -0,0 +1,252 @@
+import re
+import logging
+import os
+import requests
+
+from typing import Tuple
+from moodler.assignment import Assignment
+
+from moodler.moodle_exception import MoodlerException
+from moodler.config import TOKEN, URL
+
+logger = logging.getLogger(__name__)
+
+GRADING_WORKSHEET_UPLOAD_CONTEXTID_PATTERN = (
+    r'<input name="contextid" type="hidden" ' r'value="(\d+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_SESSKEY_PATTERN = (
+    r'<input name="sesskey" type="hidden" ' r'value="([\w\d]+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_GRADESFILE_PATTERN = (
+    r'<input type="hidden" name="gradesfile" id="id_gradesfile" ' r'value="(\d+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_IMPORTID_PATTERN = (
+    r'<input name="importid" type="hidden" ' r'value="(\d+)"'
+)
+
+GRADING_WORKSHEET_UPLOAD_DRAFTID_PATTERN = (
+    r'<input name="draftid" type="hidden" ' r'value="(\d+)"'
+)
+
+SUCCESS_GRADING_WORKSHEET_UPLOAD_PATTERN = r"Updated (\d+) grades and feedback"
+
+
+class UploadException(MoodlerException):
+    pass
+
+
+class InvalidGradingPage(UploadException):
+    pass
+
+
+class InvalidUploadGradingWorksheetPage(UploadException):
+    pass
+
+
+class InvalidUploadGradingConfirmationPage(UploadException):
+    pass
+
+
+def get_params_from_grading_page(
+    assignment_id: str, session: requests.Session
+) -> Tuple[str, str]:
+    # Build the get request to the grading page.
+    params = {"id": assignment_id, "action": "grading"}
+    response = session.get(URL + "/mod/assign/view.php", params=params)
+
+    grading_page_content = response.content.decode()
+    contextid_match = re.search(
+        GRADING_WORKSHEET_UPLOAD_CONTEXTID_PATTERN, grading_page_content
+    )
+    sesskey_match = re.search(
+        GRADING_WORKSHEET_UPLOAD_SESSKEY_PATTERN, grading_page_content
+    )
+
+    if contextid_match is None:
+        raise InvalidGradingPage(
+            "The contextid required to upload the "
+            "grading sheet to moodle was not found."
+        )
+    if sesskey_match is None:
+        raise InvalidGradingPage(
+            "The sesskey required to upload the "
+            "grading sheet to moodle was not found."
+        )
+    return contextid_match.group(1), sesskey_match.group(1)
+
+
+def get_params_from_upload_grading_worksheet_page(
```

Comment:
`get_destination_file_id`?

Reply:
I'm not sure about that. It is an HTML form parameter called `gradesfile`
maybe someone will need the same parameter in the future and he can find that this method already implemented and getting this param